CONTENTdm: metadata-only project client DIP upload

Hwesta opened the pull request:
Changes to how CONTENTdm DIP upload works, specifically for project client in a metadata-only DIP 'upload'.

simple.txt and compound.txt are now generated soley from the METS file, instead of querying the CONTENTdm server for what fields are available for this collection.  The column names are based on the XML element names of associated dmdSecs, which are generated from the incoming metadata.csv

Both simple.txt and compound.txt have the AIP UUID and file UUID added as columns.  Bulk compound.txts (only directories) have Directory name as the first column, while the rest have Filename as the last column.

Converted restructureDIPForContentDMUPload to use lxml instead of xml.dom.minidom, as lxml is more powerful for searching.  All the project client code uses lxml, but not all the direct upload code uses it, so direct upload doesn't work currently.

This relies on and currently includes changes in #147 

refs #7239

Hwesta committed:
Contentdm: METS has namespaces

refs #7239

METS files have fully qualified namespaces, instead of relying on the
default. Update contentdm to search for elements in any namespace.

Hwesta committed:
CONTENTdm: Replace project client functions with one function

Replace simple and compound project client functions with one functon that
pulls all metadata from the METS file, instead of from the server. Use
element names as column headers and element text for the row values. Prefers
non-DC metadata and falls back to DC metadata.  Adds AIP UUID and file UUID
columns to the end of the CSV.

Convert to use lxml for more powerful searching. Clean up, add docstrings,
simplify.

Hwesta committed:
CONTENTdm: Project client does not query server

refs #7239

Update MicroServiceChainLinks to not query the CONTENTdm server if the
upload type is project client.

Hwesta committed:
Restructure for CONTENTdm: General cleanup

Move comments to doc strings as appropriate. Simplify duplicate code.
Add comments with questions and ideas.

Hwesta committed:
Ingest: Make DIPs available to be downloaded

refs #7239

Add review link to the move to uploadedDIPs microservice to be able to
review and download DIPs. Consolidate directory browsing endpoint.

mistydemeo commented on src/MCPClient/lib/clientScripts/restructureDIPForContentDMUpload.py:
```diff
@@ -273,50 +270,22 @@ def getContentdmCollectionFieldInfo(contentdmServer, targetCollection):
+    Get a list of all the files (recursive) in the provided directory.
 
+    "Even though there can be subdirectories in the objects directory,
+    assumes each file should have a unique name." What does this mean?
```
I wonder if this means that filenames are (or should be) guaranteed unique across the SIP, regardless of the directory in which they're located? e.g. two subdirectories can't have objects with the same names?

Hwesta replied:
Maybe, but the paths returned in `fileList` are the full path, so that seems a weird qualifier.  Maybe it's a statement that the parts returned _are_ the full path, and are therefore unique?  It was in the original comments.

mistydemeo commented on src/MCPClient/lib/clientScripts/restructureDIPForContentDMUpload.py:
```diff
@@ -490,139 +459,127 @@ def generateFullFileEntry(title, filename, extension):
-            dmdSecPair['dc'] = parseDmdSec(dmdSecs[0])
-            dmdSecPair['nonDc'] = None
+    dmdSecPair = {'dc': None, 'nonDc': None}
+    if len(dmdSecs) > 2:
```
Use `lenDmdSecs` assigned above?

Hwesta replied:
Done

mistydemeo commented on src/MCPClient/lib/clientScripts/archivematicaCreateMETSMetadataCSV.py:
```diff
@@ -24,84 +24,103 @@
+
+    Produces:
+    {
+        'objects/foo.jpg': OrderedDict(dc.title=Foo, dc.date=2000, Other metadata=Taken on a sunny day)
```
Nitpick, but include comma at the end?

Hwesta replied:
Ooops, went in with #147

mistydemeo commented on src/MCPClient/lib/clientScripts/archivematicaCreateMETS2.py:
```diff
@@ -869,7 +860,7 @@ def create_object_metadata(struct_map):
         time.sleep(10)
 
-    parseMetadata(baseDirectoryPath)
+    CSV_METADATA = parseMetadata(baseDirectoryPath)
```
👍👍👍👍👍 🎉

mistydemeo commented:
Looks good, aside from the comments I left!

Hwesta commented:
Updated to current qa/1.x, fixed based on comments.

mistydemeo commented:
:+1:

Hwesta commented:
Holding off merging this until we can decide what to do with that currently-horribly-broken direct upload code path.

Hwesta committed:
CONTENTdm: Remove Direct Upload

Remove direct upload as an option from CONTENTdm DIPs. All CDM DIPs are
project client.

No communication with the CDM server is necessary now, so remove MSCLs and
configuration options that deal with that. Remove code in restructure DIP
that deals with direct upload.

Output simple/compound.txt to the objects directory of the DIP.

Hwesta committed:
METS: Sort structMap entries

Sort structMap entries alphabetically. Folder contents will be sorted
alphabetically within their folder. This also produces a nice
simple/compound.txt for CONTENTdm.

Hwesta commented:
Direct upload removed (MSCL, client scripts, admin config).

mistydemeo commented:
> instead of querying the CONTENTdm server for what fields are available for this collection

Out of curiosity, does this break any existing functionality?

mjordan commented:
The query to see what fields are available for a collection is comprehensive, i.e., it includes all fields that are configured for a collection. If I follow the provenance of the dmdSec values as described in the initial comment, one issue will be that if the incoming metadata.csv file doesn't contain columns corresponding to all fields for a collection (which it probably should not be trusted to do), the dmdSecs will not contain the missing fields, and consequently those fields will not get generated out into simple.txt and compound.txt.

mistydemeo commented on src/MCPClient/lib/clientScripts/restructureDIPForContentDMUpload.py:
```diff
@@ -22,614 +22,118 @@
+
+    # Iterate over all descendants and put in the return dict
+    # Key is the element's tag name, value is a list of the element's text
+    # TODO check for problems with encoding & unicode vs str
```
:scream: Here be dragons, I think...

Hwesta replied:
Checked for problems and miraculously didn't find any. If the text is unicode it writes out just fine.

mistydemeo commented on src/MCPClient/lib/clientScripts/restructureDIPForContentDMUpload.py:
```diff
@@ -22,614 +22,118 @@
 # each generateXXXProjectClientPackage function. Direct upload packages
 # have their DC metadata supplemented with the AIP UUID in generateDescFile().
 def addAipUuidToDcMetadata(dipUuid, dcMetadata):
+    if not dcMetadata:
```
What's the intent of this clause?

Hwesta replied:
If there's no DC metadata provided (None), trying to add an identifier to it will fail. This returns the falsey value as is.  Would it be clearer if it returned None?

mistydemeo replied:
I think so, yeah. I was trying to figure out if there was some semantic meaning behind the specific falsey value that was passed in.

mistydemeo commented on src/MCPClient/lib/clientScripts/restructureDIPForContentDMUpload.py:
```diff
@@ -640,569 +144,115 @@ def addAipUuidToDcMetadata(dipUuid, dcMetadata):
+    :param dmdsecs: Dict of {<DMDID>: OrderedDict{column name: value} or <dmdSec element>? }
+    :param dipuuid: UUID of the DIP
+    """
+    print 'dipuuid', dipuuid
```
Just to check, is this logging meant to be here?

Hwesta replied:
I'd planned on leaving it so I could debug things if necessary. In general, I want more output in the client scripts. I could convert it to using the logging module if you think that's better.

mistydemeo replied:
Yeah, I'm glad to have logging - I'd just rather it be easy to understand. If the output were a bit clearer/more verbose I think this'd work for me :+1:

mistydemeo commented on src/MCPClient/lib/clientScripts/restructureDIPForContentDMUpload.py:
```diff
@@ -640,569 +144,115 @@ def addAipUuidToDcMetadata(dipUuid, dcMetadata):
+
+            # Compare csv_header, if diff ERROR (first time set, write to file)
+            if csv_header_ref and csv_header_ref != csv_header:
+                print >> sys.stderr, 'ERROR headers differ,', csv_path, 'almost certainly invalid'
```
Do we want to actually stop writing, or skip the row?

Hwesta replied:
Discussed with Sarah, and looks like it should fail.  This case shouldn't happen, anyway.

mistydemeo commented:
:+1: aise from comments

Hwesta committed:
CONTENTdm: Exit with error if writing invalid tabfile

Hwesta commented:
> > instead of querying the CONTENTdm server for what fields are available for this collection
> 
> Out of curiosity, does this break any existing functionality?

Yes, but by design. Previously, this queried the CDM server to get the list of fields and performed a matching between the columns in metadata.csv to the fields in CDM.  This has been changed to read from the dmdSecs (preferring OTHER over DC), and trusting that the metadata.csv matches the columns. CDM has a matching interface to fix things up as needed, and we keep the order of the columns the same as the metadata.csv.

mistydemeo commented:
:shipit:
